bokehjs/src/lib/models/tools/inspectors/hover_tool.ts

```diff
@@ -1,81 +1,102 @@
-import {color2hex, color2css} from "core/util/color"
-import {is_empty} from "core/util/object"
+import {MoveEvent} from "core/ui_events"
+import {color2css, color2hex} from "core/util/color"
 import {enumerate} from "core/util/iterator"
-import {isString, isFunction, isNumber} from "core/util/types"
-import {build_view, build_views, remove_views} from "core/build_views"
-import {HoverMode, PointPolicy, LinePolicy, Anchor, TooltipAttachment, MutedPolicy} from "core/enums"
-import {Geometry, PointGeometry, SpanGeometry, GeometryData} from "core/geometry"
-import {ColumnarDataSource} from "../../sources/columnar_data_source"
-import {ImageIndex, Selection} from "../../selections/selection"
+import {is_empty} from "core/util/object"
+import {Formatters, FormatterType, replace_placeholders} from "core/util/templating"
+import {isFunction, isNumber, isString, is_undefined} from "core/util/types"
 import {tool_icon_hover} from "styles/icons.css"
-import {Signal} from "core/signaling"
-import {compute_renderers} from "../../util"
 import * as styles from "styles/tooltips.css"
+import {Tooltip, TooltipView} from "../../annotations/tooltip"
+import {CallbackLike1} from "../../callbacks/callback"
 import {Template, TemplateView} from "../../dom"
+import {GlyphView} from "../../glyphs/glyph"
+import {HAreaView} from "../../glyphs/harea"
+import {LineView} from "../../glyphs/line"
+import {MultiLineView} from "../../glyphs/multi_line"
+import {PatchView} from "../../glyphs/patch"
+import {VAreaView} from "../../glyphs/varea"
+import {DataRenderer} from "../../renderers/data_renderer"
+import {GlyphRenderer} from "../../renderers/glyph_renderer"
+import {GraphRenderer} from "../../renderers/graph_renderer"
+import {Renderer} from "../../renderers/renderer"
+import {ImageIndex, Selection} from "../../selections/selection"
+import {ColumnarDataSource} from "../../sources/columnar_data_source"
+import {compute_renderers} from "../../util"
+import {CustomJSHover} from "./customjs_hover"
+import {InspectTool, InspectToolView} from "./inspect_tool"
 
-export type TooltipVars = {index?: number} & Vars
-
-export function _nearest_line_hit(i: number, geometry: Geometry,
-    sx: number, sy: number, dx: Arrayable<number>, dy: Arrayable<number>): [[number, number], number] {
-  const d1 = {x: dx[i], y: dy[i]}
-  const d2 = {x: dx[i+1], y: dy[i+1]}
-
-  let dist1: number
-  let dist2: number
-  if (geometry.type == "span") {
-    if (geometry.direction == "h") {
-      dist1 = Math.abs(d1.x - sx)
-      dist2 = Math.abs(d2.x - sx)
-    } else {
-      dist1 = Math.abs(d1.y - sy)
-      dist2 = Math.abs(d2.y - sy)
+export type TooltipVars = {
+    index: number | null
+    glyph: GlyphView
+    x: number
+    y: number
+    sx: number
+    sy: number
+    tt_x: number
+    tt_y: number
+    tt_sx: number
+    tt_sy: number
+    name: string | null
+    indices?: any | undefined
+    segment_index?: any | undefined
+    image_index?: ImageIndex | undefined
```

Comment:
Those three types contain redundant `undefined` given bokehjs lib's tsconfig (it's a new and not enabled right now feature of TypeScript that allows to distinguish between `foo?: T` and `foot?: T | undefined`). Also, those `any` types seem to me easy to be filled in.

Reply:
This was following the pattern that was already there, but I will change it.

Reply:
[60556ad](https://github.com/bokeh/bokeh/pull/12147/commits/60556ad9f4b1a9a25bf12298df52dccc99ded769)